presence: Move to ids, not emails, for clients who can support them (i.e. webapp)

showell opened the pull request:
@timabbott We should definitely discuss this first.  The way this code works now is that some endpoints take slim_presence as a flag, which I think is fine, and then we also set a flag by the same name to determine how events get written to the clients.  It's possible that we should name the event flag separately from the flag that we pass in to the endpoints (although I'm mostly inclined to say they should have the same name).  And then, more importantly, it might be less code weight to just invent a new event type like 'modern_presence' or something, and then just have the webapp register for that and somehow deregister from the old 'presence'.  And then we would have less cruft in the client descriptors and no need for the 'process_presence_event' hacks.  Instead, we'd just have two calls to `send_event` when users hit presence-related endpoints, and then tornado would transmit the correct one to each client automatically.

showell commented:
Also, I modified `apply_event` so that it can handle either value for the boolean flag, but I guess I can just assume that code only runs when `slim_presence` is `True`, correct?  If so, that should be a pretty easy fix.

timabbott commented:
I don't think we can assume `apply_event` only gets `slim_presence=True`; that is part of the shared code path used by mobile `GET /register` and for generating `page_params`.

Can you add `presence:` prefixes on all the commit messages?  

For the last commit, I think we should construct the two versions of the dictionary outside the loop; it'll be a lot more efficient and I think that code path is performance-sensitive.

showell commented:
I left apply_event alone based on your response (i.e. I continue to handle both possible values of the flag).

I added "presence" prefixes to most of the commits.

I also cleaned up how we create the dictionaries in the process_presence_event code.

Finally, based on our chat conversation, I took a closer look at the code in buddy_data that helps populate the buddy list, and I made a couple cleanups there that should make you less concerned about removing suspect_online (and related refactorings that led to its removal).

timabbott commented:
Look great.  Merged as the series ending with ceb0879b0dc370c3d16c1ddc1c6119db421adf7e (added a couple pure-comment commits on top).
